ui/component/core/src/index.ts

```diff
@@ -1105,76 +962,118 @@ export class Manager implements EventProviderFactory {
-                if(this.isTokenExpired() && !this._authDisconnected) {
-                    this._tryUpdateAccessToken().then((offline) => {
-                        if(!offline) {
-                            this._setAuthDisconnected(false, true);
-                        }
-                    });
-                }
+    /**
+     * Checks keycloak is available and token is valid otherwise will redirect to login; also checks if event bus is
+     * online.
+     */
+    public async reconnect(reattemptDelayMillis: number = 3000) {
 
-            } else {
-                // When EventProvider is back online, check if authentication is disconnected and try to reauthenticate if necessary.
-                // Only emit the ONLINE event when both are connected.
-                this._eventsDisconnected = false;
-                if(this._authDisconnected) {
-                    this._runAuthReconnectTimer()
-                } else {
-                    this._emitEvent(OREvent.ONLINE);
+        if (!this._disconnected) {
+            return;
+        }
+
+        if (this._reconnectTimer) {
+            window.clearTimeout(this._reconnectTimer);
+            this._reconnectTimer = undefined;
+        }
+
+        const tryReconnect = async () => {
+            let keycloakOffline = !await this.isKeycloakReachable();
+
+            if (keycloakOffline) {
+                return false;
+            }
+
+            // Check if access token can be refreshed
+            try {
+                await this._updateKeycloakAccessToken();
+            } catch (e) {
+                // Try and use offline token if it is available
+                const offlineToken = await this._getNativeOfflineRefreshToken();
+                this._keycloak!.refreshToken = offlineToken;
+
+                try {
+                    await this._updateKeycloakAccessToken();
+                } catch (e) {
+                    this.login();
+                    return;
                 }
+                return true;
             }
+
+            // Check events
+            const eventsOffline = this.events && this.events.status === EventProviderStatus.CONNECTING;
+            // Force reconnect attempt now if needed
+            return !eventsOffline || await this.events?.connect();
+        };
+
+        const connected = await tryReconnect();
+
+        if (connected === undefined) {
+            // Going back to keycloak login so nothing to do
+            return;
         }
-    }
 
-    /** Function that clears the `WebView` history of a console. It will not delete the history on regular browsers. */
-    protected _clearWebHistory(): void {
-        this.console?._doSendGenericMessage("CLEAR_WEB_HISTORY", undefined);
+        if (!connected) {
+            // Schedule reconnect again
+            reattemptDelayMillis = Math.min(Manager.MAX_RECONNECT_DELAY, reattemptDelayMillis + 3000)
```

Comment:
Should we make those 3000ms a "global" value similar to `MAX_RECONNECT_DELAY` ?

Reply:
Could extract as a constant but not the highest priority unless it was used in multiple places